runtime/weak.c

```diff
@@ -89,7 +94,29 @@ Caml_inline int Must_be_Marked_during_mark(value x)
   CAMLassert (caml_gc_phase == Phase_mark);
   return Is_block (x) && Is_in_heap (x);
 }
+Caml_inline int Is_White_During_Mark(value x)
+{
+  CAMLassert (caml_gc_phase == Phase_mark);
+  return x != caml_ephe_none && Is_block (x) && Is_in_heap (x) && Is_white_val(x);
+}
+#endif
+
+/* True if ar could be in the set (1) (Cf. major_gc.c) */
+static int already_marked (value ar)
+{
+  value old_data;
+  CAMLassert (caml_gc_phase == Phase_mark);
+
+  old_data = Field (ar, CAML_EPHE_DATA_OFFSET);
+  return ( old_data == caml_ephe_none ||
+           (Is_block(old_data) &&
```

Comment:
Either I'm confused or you are missing a negation here.

In any case, you should define an inline function that returns
```
#if defined(NATIVE_CODE) && defined(NO_NAKED_POINTERS)
x == Is_block(x) && !Is_young(x) && Is_white_val(x)
#else
x == Is_block(x) && Is__in_heap(x) && Is_white_val(x)
#endif
```
and then use it to define `Is_white_during_mark` and `Is_dead_during_clean` (as well as this function) without the need for ugly `#ifdef`s.